libcudacxx/include/cuda/try_cancel

```diff
@@ -0,0 +1,324 @@
+ * NVIDIA, 2788 San Tomas Expressway, Santa Clara, CA 95051.
+ *
+ * 15. ENTIRE AGREEMENT. This license is the final, complete and exclusive agreement between the parties relating to the
+ * subject matter of this license and supersedes all prior or contemporaneous understandings and agreements relating to
+ * this subject matter, whether oral or written. If any court of competent jurisdiction determines that any provision of
+ * this license is illegal, invalid or unenforceable, the remaining provisions will remain in full force and effect.
+ * This license may only be modified in a writing signed by an authorized representative of each party.
+ *
+ * (v. August 20, 2021)
+ */
+
+#ifndef _CUDA_TRY_CANCEL
+#define _CUDA_TRY_CANCEL
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+namespace experimental {
+
+namespace __detail {
+
+struct __empty_t {
+  __device__ void operator()(dim3);
+};
+
+template <int __I>
+__device__ int __cluster_get_dim(__int128 __result) {
+    int __r;
+    if constexpr(__I == 0) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::x.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else if constexpr(__I == 1) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::y.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else if constexpr(__I == 2) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::z.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else {
+      __r = -1;
+      #ifndef NDEBUG
+      __trap();
+      #endif
+    }
+    return __r;
+}
+
+} // namespace __detail
+
+/// This API for implementing work-stealing, repeatedly attempts to cancel the launch of a thread block
+/// from the current grid. On success, it invokes the unary function `__uf` before trying again.
+/// On failure, it returns.
+///
+/// This API does not provide any memory synchronization.
+/// This API does not guarantee that any thread will invoke `__uf` with the next block index until all
+/// invocatons of `__uf` for the prior block index have returned.
+///
+/// Preconditions:
+/// - All thread block threads shall call this API exactly once.
+/// - Exactly one thread block thread shall call this API with `__is_leader` equals `true`.  
+template <int __ThreadBlockDim = 3, typename __UnaryFunction = __detail::__empty_t>
+    requires std::is_invocable_r_v<void, __UnaryFunction, dim3>
+__device__ void try_cancel_blocks(bool __is_leader, __UnaryFunction __uf) {
+  #if (__CUDA_ARCH__ < 1000)
+    __trap();
+  #else   // NV_HAS_FEATURE_SM100
```

Comment:
This does not work for nvc++, I am wondering whether this should rather be in the PTX exposure

Reply:
I can provide a fallback for`__CUDA_ARCH__ < 1000`. What's the right way to check for that? I did not find any docs for `NV_IF_TARGET` family of macros, so ended up using `__CUDA_ARCH__`.

Reply:
```cpp
NV_IF_ELSE_TARGET(NV_PROVIDES_SM100, (do_thething();), (error_path();))
```

Reply:
Updated and provided fallback for older SMs. Still not sure whether that's the right thing to do. Have left a TODO to resolve that before merging.